### DISCUSSION TITLE
Simple Flow Typing for Unit

### DISCUSSION THREAD

**Discussion by AugustNagro** [scala/scala3#17878] [Feature Requests]
It is really cool that Scala 3 supports [flow-sensitive typing for null][1].

It would also be helpful to have flow-typing for Unit. In Scala.js Unit maps to JavaScript undefined, and often when modeling npm library types you will have fields that may be undefined.

For example this works in Scala 3:
```Scala
val x: Int | Null = ???

if x == null then println("was null")
else println(x + 1)
```

But not this:

```Scala
val x: Int | Unit = ???

if x == () then println("was undefined")
else println(x + 1)
```

The current solution in Scala.js is to use [js.UndefOr](https://www.scala-js.org/api/scalajs-library/latest/scala/scalajs/js/index.html#UndefOr[+A]=A@scala.annotation.unchecked.uncheckedVariance|Unit), but I think it may be nicer to use Scala 3 union types.

If this is something that could be accepted with a PR I could try my hand at implementing.

[1]: https://dotty.epfl.ch/docs/reference/other-new-features/explicit-nulls.html#flow-typing

**Comment by smarter** [comment]
WDYT @sjrd?

**Comment by sjrd** [comment]
Oh I'm all for it, of course. I was already suggesting this when we started discussing flow typing for `null`, back when we were still physically talking around a table...

**Comment by AugustNagro** [comment]
I spent some time looking at this [PR](https://github.com/lampepfl/dotty/pull/7556/files), creating a class Unitables.scala & started updating all relevant places.

But it seems like a bigger refactoring & codebase knowledge than I have the bandwidth for is required. Sorry guys.
